src/ExifViewer.jl

```diff
@@ -32,75 +56,80 @@ Reads EXIF data from a filepath, IO, vector{Int8} and returns a Dict with tags a
-
-julia> a = read("test/test_images/test.jpg")
-julia> read_metadata(a)
+    "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+    "EXIF_TAG_ORIENTATION"       => "Top-left"
+    "EXIF_TAG_EXIF_VERSION"      => "Exif Version 2.1\0ion"
+    ⋮                            => ⋮
+  
+julia> read_metadata(file)
 Dict{Any, Any} with 12 entries:
-  "EXIF_TAG_DOCUMENT_NAME"     => "Test_Image"
-  "EXIF_TAG_ORIENTATION"       => "Bottom-right"
-  ⋮                            => ⋮
-
-```
-
-
-```julia
-julia> using BenchmarkTools, TestImages, ExifViewer
-julia> filename = "test/test_images/test.jpg"
-
-julia> io = open(filename, "r")
-julia> @btime read_metadata(io)
-1.862 μs (11 allocations: 912 bytes)
-
-julia> @btime read_metadata(filename)
-269.472 μs (122 allocations: 100.05 KiB)
-
-julia> @btime data = read(filename)
-15.876 μs (13 allocations: 69.28 KiB)
-
-julia> @btime read_metadata(data)
-221.786 μs (109 allocations: 30.77 KiB)
+      "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+      "EXIF_TAG_ORIENTATION"       => "Top-left"
+      "EXIF_TAG_EXIF_VERSION"      => "Exif Version 2.1\0ion"
+      ⋮                            => ⋮
+    
 ```
 """
-function read_metadata(data::Vector{UInt8}; ifds::Union{Int, NTuple, UnitRange}=1)
+function read_metadata(
+    data::Vector{UInt8};
+    allifds = true,
+    ifds::Union{Int,NTuple,UnitRange} = 1,
+    extract_thumbnail = false,
+)
     ed_ptr = LibExif.exif_data_new_from_data(data, length(data))
-    # handle case where ed_ptr is null
-    if(ed_ptr == C_NULL)
-        return error("Unable to read EXIF data")
+    if (ed_ptr == C_NULL)
+        return error("Unable to read EXIF data: invalid pointer")
     end
     ed = unsafe_load(ed_ptr)
-    ifds = collect(1:numifds(ed)) # reads all
-    # ifds = collect(ifds) # assumes user knows what they are doing
-    result = Dict()
+
+    if (allifds == true)
+        ifds = collect(1:numifds(ed))
+    else
+        ifds = collect(ifds)
+    end
+
+    result = Dict{String,Any}()
+    str = Vector{Cuchar}(undef, 1024)
     for i in ifds
         content_ptr = ed.ifd[i] # ques: do we need to unref these too?
         # handle case where content_ptr is null
-        if(content_ptr == C_NULL)
-            return error("Unable to read IFD:",i)
+        if (content_ptr == C_NULL)
+            return error("Unable to read IFD:", i)
         end
         data = unsafe_load(content_ptr)
+        if data.count == 0
+            continue
+        end
         res = unsafe_wrap(Array, data.entries, data.count)
-        for i in 1:data.count
-            str = Vector{Cuchar}(undef, 1024);
-            LibExif.exif_entry_get_value(Ref(unsafe_load(res[i])), str, length(str))
-            result[string(unsafe_load(res[i]).tag)] = rstrip(String(str), '\0')
+        for i = 1:data.count
+            entry = unsafe_load(res[i])
+            LibExif.exif_entry_get_value(Ref(entry), str, length(str))
+            tag = String(copy(str))[1:findfirst(iszero, str)-1]
```

Comment:
Ah, I see. Maybe we can use `StringView` here to tweak the allocation a bit? This is because both `copy` and getindex `[1:findfirst(iszero, str)-1]` has a copy(allocation) semantic.

```julia
julia> A = UInt8['h', 'e', 'l', 'l', 'o', '\0']
6-element Vector{UInt8}:
 0x68
 0x65
 0x6c
 0x6c
 0x6f
 0x00

julia> B = StringView(A)[1:length(A)-1]
"hello"

julia> A
6-element Vector{UInt8}:
 0x68
 0x65
 0x6c
 0x6c
 0x6f
 0x00
```

Reply:
Quite tempted to use it(looks lot cleaner) but its 1 extra package for a single line of code, given read_metadata and read_tags are likely to be merged:). I am interested in the memory allocations happening quite a lot so will see if this works better for us.

Reply:
This behavior or stringviews.jl is interesting frankly:
```
function read_tags(
    data::Vector{UInt8};
    ifds::Union{Int,NTuple,UnitRange} = 1:5,
    read_all = false,
    tags::Vector{LibExif.ExifTag} = Vector{LibExif.ExifTag}([]),
    extract_thumbnail = false,
)
    ed_ptr = LibExif.exif_data_new_from_data(data, length(data))
    if (ed_ptr == C_NULL)
        return error("Unable to read EXIF data: invalid pointer")
    end

    ed = unsafe_load(ed_ptr)

    ifds = collect(ifds)
    checkbounds(Bool, collect(1:numifds(ed)), ifds) || throw(BoundsError(collect(1:numifds(ed)), ifds))

    tags = read_all ? tags : Set(tags)
    result = Dict{String,Any}()
    str = Vector{Cuchar}(undef, 1024)

    for i in ifds
        content_ptr = ed.ifd[i] # ques: do we need to unref these too?
        if (content_ptr == C_NULL)
            return error("Unable to read IFD:", i)
        end
        data = unsafe_load(content_ptr)
        if data.count == 0
            continue
        end
        res = unsafe_wrap(Array, data.entries, data.count)
        for i = 1:data.count
            entry = unsafe_load(res[i])
            condition = read_all ? read_all : entry.tag in tags
            if condition
                # println("Now:",StringView(str[1:50]))
                LibExif.exif_entry_get_value(Ref(entry), str, length(str))
                # println("After:",StringView(str[1:50]))
                # tag = String(copy(str))[1:findfirst(iszero, str)-1]
                # tag = StringView(str)[1:findfirst(iszero, str)-1]
                # tag = StringView(str)[1:length(str)-1]
                tag = StringView(str)[1:findfirst(iszero, str)-1]
                # print(tag)
                # print(str)
                tag = fixformat(tag, entry.format)
                # println(tag)
                if string(entry.tag) ∉ keys(result)
                    println("Tag:", string(entry.tag),", Value:", tag)
                    println(typeof(tag))
                    result[string(entry.tag)] = tag
                end

                println(result,"\n\n")
            end
            if read_all == false
                delete!(tags, entry.tag)
                if tags == Set() break end
            end
        end
    end

    # not sure we should include this
    # if isempty(tags) != true
    #     @info "Non-Existent Tags:" tags
    # end

    if (extract_thumbnail == true)
        thumbnail_size = Int(ed.size)
        thumbnail_data = unsafe_wrap(Array, ed.data, thumbnail_size)
        result["EXIF_TAG_THUMBNAIL_DATA"] = thumbnail_data
    end

    LibExif.exif_data_unref(ed_ptr)
    return result
end
```

Result:
```
julia> @time read_metadata(filepath)
Tag:EXIF_TAG_ORIENTATION, Value:Top-left
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_ORIENTATION" => "Top-left")


Tag:EXIF_TAG_X_RESOLUTION, Value:300
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_ORIENTATION" => "300\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "300")


Tag:EXIF_TAG_Y_RESOLUTION, Value:300
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_Y_RESOLUTION" => "300", "EXIF_TAG_ORIENTATION" => "300\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "300")


Tag:EXIF_TAG_RESOLUTION_UNIT, Value:Inch
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_Y_RESOLUTION" => "Inc", "EXIF_TAG_ORIENTATION" => "Inch\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "Inch", "EXIF_TAG_X_RESOLUTION" => "Inc")


Tag:EXIF_TAG_SOFTWARE, Value:Adobe Photoshop CS Windows
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_SOFTWARE" => "Adobe Photoshop CS Windows", "EXIF_TAG_Y_RESOLUTION" => "Ado", "EXIF_TAG_ORIENTATION" => "Adobe Ph", "EXIF_TAG_RESOLUTION_UNIT" => "Adob", "EXIF_TAG_X_RESOLUTION" => "Ado")


Tag:EXIF_TAG_DATE_TIME, Value:2005:05:13 13:09:04
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_SOFTWARE" => "2005:05:13 13:09:04\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "200", "EXIF_TAG_ORIENTATION" => "2005:05:", "EXIF_TAG_DATE_TIME" => "2005:05:13 13:09:04", "EXIF_TAG_RESOLUTION_UNIT" => "2005", "EXIF_TAG_X_RESOLUTION" => "200")


Tag:EXIF_TAG_COMPRESSION, Value:JPEG compression
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_SOFTWARE" => "JPEG compression\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "JPE", "EXIF_TAG_ORIENTATION" => "JPEG com", "EXIF_TAG_DATE_TIME" => "JPEG compression\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "JPEG", "EXIF_TAG_COMPRESSION" => "JPEG compression", "EXIF_TAG_X_RESOLUTION" => "JPE")


Dict{String, Any}("EXIF_TAG_SOFTWARE" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "72\0", "EXIF_TAG_ORIENTATION" => "72\0\0\0\0\0\0", "EXIF_TAG_DATE_TIME" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "72\0\0", "EXIF_TAG_COMPRESSION" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "72\0")


Dict{String, Any}("EXIF_TAG_SOFTWARE" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "72\0", "EXIF_TAG_ORIENTATION" => "72\0\0\0\0\0\0", "EXIF_TAG_DATE_TIME" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "72\0\0", "EXIF_TAG_COMPRESSION" => "72\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "72\0")


Dict{String, Any}("EXIF_TAG_SOFTWARE" => "Inch\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "Inc", "EXIF_TAG_ORIENTATION" => "Inch\0\0\0\0", "EXIF_TAG_DATE_TIME" => "Inch\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "Inch", "EXIF_TAG_COMPRESSION" => "Inch\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "Inc")


Tag:EXIF_TAG_COLOR_SPACE, Value:sRGB
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_COLOR_SPACE" => "sRGB", "EXIF_TAG_SOFTWARE" => "sRGB\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "sRG", "EXIF_TAG_ORIENTATION" => "sRGB\0\0\0\0", "EXIF_TAG_DATE_TIME" => "sRGB\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "sRGB", "EXIF_TAG_COMPRESSION" => "sRGB\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "sRG")


Tag:EXIF_TAG_PIXEL_X_DIMENSION, Value:3000
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_COLOR_SPACE" => "3000", "EXIF_TAG_SOFTWARE" => "3000\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "300", "EXIF_TAG_ORIENTATION" => "3000\0\0\0\0", "EXIF_TAG_DATE_TIME" => "3000\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_RESOLUTION_UNIT" => "3000", "EXIF_TAG_COMPRESSION" => "3000\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "300", "EXIF_TAG_PIXEL_X_DIMENSION" => "3000")


Tag:EXIF_TAG_PIXEL_Y_DIMENSION, Value:3002
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_COLOR_SPACE" => "3002", "EXIF_TAG_SOFTWARE" => "3002\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "300", "EXIF_TAG_ORIENTATION" => "3002\0\0\0\0", "EXIF_TAG_DATE_TIME" => "3002\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_PIXEL_Y_DIMENSION" => "3002", "EXIF_TAG_RESOLUTION_UNIT" => "3002", "EXIF_TAG_COMPRESSION" => "3002\0\0\0\0\0\0\0\0\0\0\0\0", "EXIF_TAG_X_RESOLUTION" => "300", "EXIF_TAG_PIXEL_X_DIMENSION" => "3002")


Tag:EXIF_TAG_EXIF_VERSION, Value:Exif Version 2.1
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_ORIENTATION" => "Exif Ver", "EXIF_TAG_EXIF_VERSION" => "Exif Version 2.1", "EXIF_TAG_RESOLUTION_UNIT" => "Exif", "EXIF_TAG_COMPRESSION" => "Exif Version 2.1", "EXIF_TAG_PIXEL_X_DIMENSION" => "Exif", "EXIF_TAG_COLOR_SPACE" => "Exif", "EXIF_TAG_SOFTWARE" => "Exif Version 2.1\0ion\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "Exi", "EXIF_TAG_PIXEL_Y_DIMENSION" => "Exif", "EXIF_TAG_DATE_TIME" => "Exif Version 2.1\0io", "EXIF_TAG_X_RESOLUTION" => "Exi")


Tag:EXIF_TAG_FLASH_PIX_VERSION, Value:FlashPix Version 1.0
StringViews.StringView{SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}}
Dict{String, Any}("EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0", "EXIF_TAG_ORIENTATION" => "FlashPix", "EXIF_TAG_EXIF_VERSION" => "FlashPix Version", "EXIF_TAG_RESOLUTION_UNIT" => "Flas", "EXIF_TAG_COMPRESSION" => "FlashPix Version", "EXIF_TAG_PIXEL_X_DIMENSION" => "Flas", "EXIF_TAG_COLOR_SPACE" => "Flas", "EXIF_TAG_SOFTWARE" => "FlashPix Version 1.0\0\0\0\0\0\0", "EXIF_TAG_Y_RESOLUTION" => "Fla", "EXIF_TAG_PIXEL_Y_DIMENSION" => "Flas", "EXIF_TAG_DATE_TIME" => "FlashPix Version 1.", "EXIF_TAG_X_RESOLUTION" => "Fla")


  0.841485 seconds (726.09 k allocations: 45.952 MiB, 1.18% gc time, 85.92% compilation time)
Dict{String, Any} with 12 entries:
  "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
  "EXIF_TAG_ORIENTATION"       => "FlashPix"
  "EXIF_TAG_EXIF_VERSION"      => "FlashPix Version"
  "EXIF_TAG_RESOLUTION_UNIT"   => "Flas"
  "EXIF_TAG_COMPRESSION"       => "FlashPix Version"
  "EXIF_TAG_PIXEL_X_DIMENSION" => "Flas"
  "EXIF_TAG_COLOR_SPACE"       => "Flas"
  "EXIF_TAG_SOFTWARE"          => "FlashPix Version 1.0\0\0\0\0\0\0"
  "EXIF_TAG_Y_RESOLUTION"      => "Fla"
  "EXIF_TAG_PIXEL_Y_DIMENSION" => "Flas"
  "EXIF_TAG_DATE_TIME"         => "FlashPix Version 1."
  "EXIF_TAG_X_RESOLUTION"      => "Fla"
```

Reply:
Allocations does decrease, but not sure if it adds a lot of value
```
julia> @time read_metadata(filepath) # with current
  0.003571 seconds (209 allocations: 6.252 MiB)
Dict{String, Any} with 12 entries:
  "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
  "EXIF_TAG_ORIENTATION"       => "Top-left"
  "EXIF_TAG_EXIF_VERSION"      => "Exif Version 2.1"
  "EXIF_TAG_RESOLUTION_UNIT"   => "Inch"
  "EXIF_TAG_COMPRESSION"       => "JPEG compression"

julia> @time read_metadata(filepath) # with stringview version
  0.003398 seconds (191 allocations: 6.221 MiB)
Dict{String, Any} with 12 entries:
  "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\…
  "EXIF_TAG_ORIENTATION"       => "FlashPix Version 1.0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\…
  "EXIF_TAG_EXIF_VERSION"      => "FlashPix Version 1.0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\…
  "EXIF_TAG_RESOLUTION_UNIT"   => "FlashPix Version 1.0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\…
  "EXIF_TAG_COMPRESSION"       => "FlashPix Version 1.0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\…
```